src/Mod/Draft/draftmake/make_copy.py

```diff
@@ -61,152 +44,40 @@ def make_copy(obj, force=None, reparent=False):
-
-    elif (utils.get_type(obj) == "Structure") or (force == "Structure"):
-        import ArchStructure
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchStructure._Structure(newobj)
-        if App.GuiUp:
-            ArchStructure._ViewProviderStructure(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Wall") or (force == "Wall"):
-        import ArchWall
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchWall._Wall(newobj)
-        if App.GuiUp:
-            ArchWall._ViewProviderWall(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Window") or (force == "Window"):
-        import ArchWindow
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchWindow._Window(newobj)
-        if App.GuiUp:
-            ArchWindow._ViewProviderWindow(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Panel") or (force == "Panel"):
-        import ArchPanel
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchPanel._Panel(newobj)
-        if App.GuiUp:
-            ArchPanel._ViewProviderPanel(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Sketch") or (force == "Sketch"):
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject("Sketcher::SketchObject", _name)
-        for geo in obj.Geometry:
-            newobj.addGeometry(geo)
-        for con in obj.Constraints:
-            newobj.addConstraint(con)
-
-    elif hasattr(obj, 'Shape'):
+    if simple_copy and hasattr(obj, 'Shape'):
         _name = utils.get_real_name(obj.Name)
         newobj = App.ActiveDocument.addObject("Part::Feature", _name)
         newobj.Shape = obj.Shape
+        gui_utils.format_object(newobj, obj)
+    elif not simple_copy and hasattr(obj, 'Shape'):
+        newobj = App.ActiveDocument.copyObject(obj)
 
-    else:
-        print("Error: Object type cannot be copied")
+    if not newobj:
         return None
 
-    for p in obj.PropertiesList:
-        if not p in ["Proxy", "ExpressionEngine"]:
-            if p in newobj.PropertiesList:
-                if not "ReadOnly" in newobj.getEditorMode(p):
-                    try:
-                        setattr(newobj, p, obj.getPropertyByName(p))
-                    except AttributeError:
-                        try:
-                            setattr(newobj, p, obj.getPropertyByName(p).Value)
-                        except AttributeError:
-                            pass
-
     if reparent:
         parents = obj.InList
         if parents:
             for par in parents:
                 if par.isDerivedFrom("App::DocumentObjectGroup"):
                     par.addObject(newobj)
-                else:
+                else: # Carlo: when is it used?
```

Comment:
This `else` is executed when the `par` is not a `Std_Group`. There may be other objects which add objects inside its `InList` which aren't necessarily groups, like Walls maybe?

Reply:
mmm... could be, but i'm not sure... perhaps yorik can shed more light... I was thinking to App::Part instead...

Reply:
That I remember, `InList` and `OutList` aren't modified directly; these are set automatically in the case of groups, when you `addObject` things. Make a test, create one group, then create another group and place it inside the first one, then a third group and nest it inside the second one. Then check the values of `InList` and `OutList`. You will hopefully get an idea of how these attributes are set.
```
InList
|
|--- group
      |
      |---- OutList
```

So, what I mentioned before, maybe a Wall would be in the InList of a Window. And maybe the Window would be in the OutList of the Wall.

Reply:
Yes, you are right, but it's still not so clear to me how it is supposed to work: for which property and for which objects.
@yorikvanhavre, can you shed some light?

Reply:
I don't remember exactly, but for example Arch_BuildingParts or Draft_Layers could be in that case... I would maybe just scan for any parent with a "Group" property, and if the original object is in the Group, then the copy goes too... That would I think cover all the needed cases and leave out the cases where we don't want that (ie. Walls and such).

Reply:
Tested this code with layers:
`                else:
                    # That's the case of Arch_BuildingParts or Draft_Layers for example
                    if "Group" in par.PropertiesList:
                        if obj in par.Group:
                            group = par.Group
                            group.append(newobj)
                            par.Group = group
`

Seems to do the trick for copy with move, while copy with rotate does not work... I guess I have to refactor it a bit as I did with Move.